File: jquery_/trunk/doc/devtools/mergelang.pl
#!/usr/bin/perl
# A problem with having to maintain translations of several versions of tiki is that the same strings
# have the same translation, but is not desirable to change twice the same strings
# This utility takes the translations for the old version, and if the string is in the new translation file
# it puts the old translation there.


if ($#ARGV==1) {
	$fileold=$ARGV[0];
	$filenew=$ARGV[1];

	open OLD, "<$fileold";
	while ($line=<OLD>) {
		if ($line=~/^\s*\"(.*)\"\s*\=\>\s*\"(.*)\"\s*\,\s*$/) {
			$orig=$1;
			$trad=$2;
			if ($orig ne $trad) {
				$trads{$orig}=$trad;
			}
		}
	}
	close OLD;
	
	open CURRENT,"<$filenew";
	open NEW,">$filenew.new";
	while ($line=<CURRENT>) {
		if ($line=~/^\s*\"(.*)\"\s*\=\>\s*\"(.*)\"\s*\,\s*$/) {
			$orig=$1;
			$trad=$2;
			if ($trads{$orig}) {
				print NEW "\"$orig\" => \"" . $trads{$orig} . "\",\n";
			} else {
				print NEW "\"$orig\" => \"$trad\",\n";
			}
		} else {
			print NEW $line;
		}
	}
	close CURRENT;
	close NEW;
} else {
	print $#ARGV . "\n";
	print "Usage: mergelang.pl oldlang.php currentlang.php\n";
	print "This will generate a currentlang.php.new file with the content of the currentlang.php file but with the old translations for the matching strings from oldlang.php\n";
}

File: doc/devtools/mergelang.php
<?php
// (c) Copyright 2002-2011 by authors of the Tiki Wiki CMS Groupware Project
// 
// All Rights Reserved. See copyright.txt for details and a complete list of authors.
// Licensed under the GNU LESSER GENERAL PUBLIC LICENSE. See license.txt for details.
// $Id$

/*
 *  A problem with having to maintain translations of several versions of Tiki is
 * that the same strings have the same translations, but is not desirable to change
 * them for each branch. This utility takes the translations from a source version
 * and add to a target version if the string exists in the targe language files. 
 */

if ($argc < 3) {
	$helpMsg = "\nUsage: php doc/devtools/mergelang.php pathToSourceTikiVersion pathToTargetTikiVersion [lang=pt-br,es]\n";
	$helpMsg .= "\nExamples: \n\t\tphp doc/devtools/mergelang.php /home/user/devel/tiki80 /home/user/deve/tiki60";
	$helpMsg .= "\n\t\tphp doc/devtools/mergelang.php /home/user/devel/tiki80 /home/user/deve/tiki60 lang=pt-br,es\n\n"; 
	die($helpMsg);
}

require_once('lib/language/Language.php');

$sourcePath = $argv[1];
$targetPath = $argv[2];

if (isset($argv[3])) {
	list($key, $value) = explode('=', $argv[3]);
	$languages = explode(',', $value);
} else {
	$languages = Language::getLanguages();
}

if (!file_exists($sourcePath)) {
	die("\nPath $sourcePath does not exist.\n\n");
}

if (!file_exists($targetPath)) {
	die("\nPath $targetPath does not exist.\n\n");
}

foreach ($languages as $language) {
	$sourceLangFile = "$sourcePath/lang/$language/language.php";
	$targetLangFile = "$targetPath/lang/$language/language.php";
	$tmpTargetLangFile = "$targetPath/lang/$language/language.php.tmp"; 
	
	if (file_exists($sourceLangFile)) {
		$lines = file($sourceLangFile);
		$sourceTranslations = array(); 
		
		foreach ($lines as $line) {
			$matches = array();
			
			// build an array with all the translations from the source file
			if (preg_match('/^\s*\"(.*)\"\s*\=\>\s*\"(.*)\"\s*\,\s*$/', $line, $matches)) {
				$sourceTranslations[$matches[1]] = $matches[2];
			}
		}
	}
	
	if (file_exists($targetLangFile)) {
		$lines = file($targetLangFile);
		$handle = fopen("$targetLangFile.tmp", 'w');
		
		if ($handle) {
			// foreach each line in the target file check decide whether to keep the
			// current translation or use the translation from the source file if one exists 
			foreach ($lines as $line) {
				$matches = array();
				
				if (preg_match('|^/?/?\s*\"(.*)\"\s*\=\>\s*\"(.*)\"\s*\,\s*$|', $line, $matches)
					&& isset($sourceTranslations[$matches[1]]))
				{
					fwrite($handle, "\"{$matches[1]}\" => \"{$sourceTranslations[$matches[1]]}\",\n");
				} else {
					fwrite($handle, $line);
				}
			}
			
			rename($tmpTargetLangFile, $targetLangFile);
		}
	}
}